### PR TITLE
[SR-6826] Use the remaining half bit in the refcount to bypass ObjC deallocation overhead

### PR DESCRIPTION
**Pull request by Catfish-Man**
I have no idea how to reference a JIRA issue in GitHub apparently

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/SwiftShims/RuntimeShims.h [hunk 1]
```diff
@@ -48,6 +48,9 @@ const char *_swift_stdlib_strtof_clocale(const char *nptr, float *outResult);
 SWIFT_RUNTIME_STDLIB_API
 void _swift_stdlib_immortalize(void *obj);
   
+SWIFT_RUNTIME_STDLIB_API
+void _swift_stdlib_set_objc_complications_forbidden(void *obj, int forbid);
```
Comment (lorentey) [inline stdlib/public/SwiftShims/RuntimeShims.h]: Do we have a way to mark availability of runtime entry points?

File: stdlib/public/core/ContiguousArrayBuffer.swift [hunk 2]
```diff
@@ -191,8 +191,9 @@ internal struct _ContiguousArrayBuffer<Element> : _ArrayBufferProtocol {
       _storage = Builtin.allocWithTailElems_1(
          _ContiguousArrayStorage<Element>.self,
          realMinimumCapacity._builtinWordValue, Element.self)
-
+      
       let storageAddr = UnsafeMutableRawPointer(Builtin.bridgeToRawPointer(_storage))
+      _swift_stdlib_set_objc_complications_forbidden(storageAddr, 1)
```
Comment (lorentey) [inline stdlib/public/core/ContiguousArrayBuffer.swift]: This will still need an availability check 😔

**Comment by Catfish-Man** [comment]
Nice. That’s a decent starting place. Next up:

 * Dictionary/String/Set backing stores
 * Set up a way to initialize them already with the bit set instead of paying for an extra call
 * Probably want some more tests

**Review by lorentey — APPROVED** [review]
Neat! 👏

In cases where the allocation code is inlinable, we'll need to add an availability check for the new runtime function.

File: stdlib/public/SwiftShims/RuntimeShims.h (hunk 1 above)
Comment (Catfish-Man) [reply stdlib/public/SwiftShims/RuntimeShims.h]: I have no idea! I hope so

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 2 above)
Comment (Catfish-Man) [reply stdlib/public/core/ContiguousArrayBuffer.swift]: Yeah sadly so :( I wonder if there's anywhere we tail alloc stuff that we DON'T want to do this… if there's not, we could fold it into the builtin maybe? Anyway, availability check isn't the worst thing probably.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 2 above)
Comment (lorentey) [reply stdlib/public/core/ContiguousArrayBuffer.swift]: `ManagedObject` uses the same builtin, and it's probably not a good idea to set the bit on that.

Although... wild idea: what if we set this flag all non-@objc classes?

**Comment by Catfish-Man** [comment]
That's weird. I was expecting more change from adding Dictionary and String

File: stdlib/public/SwiftShims/RuntimeShims.h (hunk 1 above)
Comment (Azoy) [reply stdlib/public/SwiftShims/RuntimeShims.h]: Couldn't you do something like:
```swift
@_silgen_name("_swift_stdlib_set_objc_complications_forbidden")
@available(/*availability stuff here*/)
internal func _set_objc_complications_forbidden(_ addr: UnsafeMutableRawPointer, _ forbid: Int)
```
to give this function availability? (You can rename function name and parameters and access control, just giving an example).

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/ContiguousArrayBuffer.swift (hunk 2 above)
Comment (jrose-apple) [reply stdlib/public/core/ContiguousArrayBuffer.swift]: That needs a proposal at the very least.

**Comment by Catfish-Man** [comment]
…doh

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/SwiftShims/RefCount.h [hunk 3]
```diff
@@ -370,22 +378,36 @@ class RefCountBitsT {
...
   
   LLVM_ATTRIBUTE_ALWAYS_INLINE
   void setIsImmortal(bool value) {
-    setField(IsImmortal, value);
+    setField(IsImmortalOrFastDealloc, value);
     setField(UseSlowRC, value);
   }
   
+  LLVM_ATTRIBUTE_ALWAYS_INLINE
+  bool hasNoObjCComplications() const {
```
Comment (gottesmm) [inline stdlib/public/SwiftShims/RefCount.h]: Nice name.

**Comment by Catfish-Man** [comment]
Ok, sweet, the maximum available win here is really nice :) this is going to require libobjc changes to get though

**Comment by Catfish-Man** [comment]
Profiling suggests this should be a ~2.5% improvement on JSONDecode, which is why it didn't show up in the CI benchmark results (they filter out everything under 7%). Still happy with that; for scale, all calls to objc_retain() add up to 2.6%, it just goes to show it's hard to move the needle on macro benchmarks.

**Comment by Catfish-Man** [comment]
Not using this PR for this, since doing it properly requires ObjC changes

**Pull request closed without merging**